### DISCUSSION TITLE
12. 4 way handshake에서 서버가 패킷의 지연을 어떻게 대처하나요?

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#16] [General]
“서버에서 FIN을 전송하기 전에 전송한 패킷이, 라우팅 지연이나 패킷 유실로 인한 재전송 등으로 인해 FIN 패킷보다 늦게 도착한다면?”

**[키워드]**
- 4-way handshake
- TIME-WAIT 상태

**Comment by cutehumanS2** [comment]
4-way handshake에서 클라이언트는 서버가 연결 종료 요청(FIN) 세그먼트를 보낸 후에도, 지연된 패킷이 도착할 가능성을 대비해 일정 시간 동안 **TIME-WAIT** 상태를 유지합니다. 이를 통해 지연된 FIN 패킷이나 재전송된 FIN을 올바르게 처리하여 정상적으로 연결이 종료되도록 보장합니다.

**Comment by cutehumanS2** [comment]
**[TCP 연결 종료 상태]**
![image](https://github.com/user-attachments/assets/aadf1971-61ca-46a4-b55c-217c2067ce61)
- **FIN_WAIT-1**: 일반적인 TCP 연결 종료 과정에 있어 FIN_WAIT-1은 연결 종료의 첫 단계가 됩니다. FIN 세그먼트로서 연결 종료 요청을 보낸 액티브 클로즈 호스트는 FIN_WAIT-1 상태로 접어들게 됩니다.
- **CLOSE-WAIT**: 종료 요청인 FIN 세그먼트를 받은 패시브 클로즈 호스트가 그에 대한 응답으로 ACK 세그먼트를 보낸 후 대기하는 상태입니다.
- **FIN-WAIT-2**: FIN_WAIT-1 상태에서 ACK 세그먼트를 받게 되면 FIN_WAIT-2 상태가 됩니다. 상대 호스트의 FIN 세그먼트를 기다리는 상태입니다.
- **LAST-ACK**: CLOSE-WAIT 상태에서 FIN 세그먼트를 전송한 뒤 이에 대한 ACK 세그먼트를 기다리는 상태입니다.
- **TIME-WAIT**: 액티브 클로즈 호스트가 FIN 세그먼트를 수신한 뒤, 이에 대한 ACK 세그먼트를 전송한 뒤 접어드는 상태입니다. 패시브 클로즈 호스트가 마지막 ACK 세그먼트를 수신하면 CLOSED 상태로 전이하는 반면, TIME-WAIT 상태에 접어든 액티브 클로즈 호스트는 일정 시간을 기다린 뒤 CLOSED 상태로 전이합니다.

<br>

**[TIME-WAIT 상태가 필요한 이유는 무엇인가요?]**
: 왜 굳이 일정 시간을 기다렸다가 연결을 종료해야 하는 것일까요?
- **마지막 ACK의 손실 가능성 처리:** 가장 주요한 이유는 상대 호스트가 받았어야 할 마지막 ACK 세그먼트가 올바르게 전송되지 않았을 수 있기 때문입니다.
- **패킷 혼란 방지:** 동일한 IP 주소와 포트를 사용하는 새 연결이 빠르게 생성될 경우, 이전 연결에서 지연된 패킷이 새 연결에 혼란을 일으킬 수 있기 때문입니다.

<br>

**[추가 질문]**
- **왜 TIME-WAIT 상태는 클라이언트가 유지하나요?**
    TCP 연결 종료에서 마지막 ACK를 보내는 역할이 클라이언트에게 있기 때문입니다.
    
- **TIME-WAIT 지속 시간은 어떻게 되나요?**
    일반적으로 MSL(Maximum Segment Lifetime, 세그먼트 최대 수명)의 2배 동안 지속됩니다.

<br><br>

[참고한 자료]
<혼자 공부하는 네트워크> 04-2. TCP와 UDP 中 “TCP 상태”
